services/terrain/src/terrain/persistence/search.clj

```diff
@@ -110,46 +110,55 @@
       (throw+ es-uninitialized))))
 
 
-(defn- tags-access-filter
-  [tags memberships]
-  (letfn [(tag-filter [tag] (query/term :id {:type  "tag"
-                                             :id    tag
-                                             :path  "targets.id"
-                                             :cache false}))]
-    (query/bool :must   (query/nested :path   "userPermissions"
-                                      :filter (query/term "userPermissions.user" memberships))
-                :should (map tag-filter tags))))
+(defn- mk-filter
+  [in-folders tags memberships]
+  (let [filter-path (fn [path] (query/prefix "path" (if (= \/ (last path))
+                                                      path
+                                                      (str path \/))))
+        filter-tag  (fn [tag] (query/term :id {:type  "tag"
+                                               :id    tag
+                                               :path  "targets.id"
+                                               :cache false}))
+        perm-filter (query/nested :path   "userPermissions"
+                                  :filter (query/term "userPermissions.user" memberships))]
+    (query/bool :must   (query/bool :must perm-filter :should (map filter-path in-folders))
+                :should (map filter-tag tags))))
```

Comment:
I think I'm misunderstanding, but why is the folder path a `should` query? If I'm reading [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html) correctly, this means results could still be returned from bad paths if they match the tag filter.

Relatedly, where (if anywhere) do we set the [minimum_should_match](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-minimum-should-match.html) parameter? I guess since we have a `must` query it won't default to needing minimum-of-one, but I'd still be curious if this is just for scoring or if we require some to match.